src/go/pt-galera-log-explainer/internal.go

```
@@ -0,0 +1,218 @@
+		// special case
+		// I'm not adding pxcoperator map the same way others are used, because they do not have the same formats and same place
+		// it needs to be put on the front so that it's not 'merged' with the '{"log":"' json prefix
+		// this is to keep things as close as '^' as possible to keep doing prefix searches
+		grepRegex += "((" + strings.Join(regex.PXCOperatorMap.Compile(), "|") + ")|^{\"log\":\""
+		regexes.Merge(regex.PXCOperatorMap)
+	}
+	if CLI.Since != nil {
+		grepRegex += "(" + regex.BetweenDateRegex(CLI.Since, CLI.PxcOperator) + "|" + regex.NoDatesRegex(CLI.PxcOperator) + ")"
+	}
+	grepRegex += ".*"
+	grepRegex += "(" + strings.Join(regexToSendSlice, "|") + ")"
+	if CLI.PxcOperator {
+		grepRegex += ")"
+	}
+	logger.Debug().Str("grepArg", grepRegex).Msg("compiled grep arguments")
+	return grepRegex
+}
+
+func execGrepAndIterate(path, compiledRegex string, stdout chan<- string) error {
+
+	defer close(stdout)
+
+	// A first pass is done, with every regexes we want compiled in a single one.
+
+	/*
+		Regular grep is actually used
+
+		There are no great alternatives, even less as golang libraries.
+		grep itself do not have great alternatives: they are less performant for common use-cases, or are not easily portable, or are costlier to execute.
+		grep is everywhere, grep is good enough, it even enable to use the stdout pipe.
+
+		The usual bottleneck with grep is that it is single-threaded, but we actually benefit
+		from a sequential scan here as we will rely on the log order.
+
+		Also, being sequential also ensure this program is light enough to run without too much impacts
+		It also helps to be transparent and not provide an obscure tool that work as a blackbox
+	*/
+	if runtime.GOOS == "darwin" && CLI.GrepCmd == "grep" {
+		logger.Warn().Msg("On Darwin systems, use 'pt-galera-log-explainer --grep-cmd=ggrep' as it requires grep v3")
+	}
+
+	cmd := exec.Command(CLI.GrepCmd, "-P", compiledRegex, path)
+
+	out, _ := cmd.StdoutPipe()
```

Better to not ignore errors, it can save a lot of time later.

---

Agreed, my mistake